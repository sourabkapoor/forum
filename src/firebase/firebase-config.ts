// Import the functions you need from the SDKs you need

import { initializeApp } from "firebase/app";
import { getStorage } from "firebase/storage";

// TODO: Add SDKs for Firebase products that you want to use

// https://firebase.google.com/docs/web/setup#available-libraries

// Your web app's Firebase configuration

const firebaseConfig = {
  // apiKey: "AIzaSyCLwXf77NuoR0ESLwAAgez_wS_gM8V4J7M",

  // authDomain: "reddit-clone-23782.firebaseapp.com",

  // projectId: "reddit-clone-23782",

  // storageBucket: "reddit-clone-23782.appspot.com",

  // messagingSenderId: "855149209259",

  // appId: "1:855149209259:web:cb03f2ae83010e6feab9ee",

  // measurementId: "G-9D511367MB",
  apiKey: "AIzaSyCEQgSH4Vfzmbn6IVFOE5fu5DtH0p3pAZ8",

  authDomain: "forum-ovidiu-borze.firebaseapp.com",

  projectId: "forum-ovidiu-borze",

  storageBucket: "forum-ovidiu-borze.appspot.com",

  messagingSenderId: "555287825535",

  appId: "1:555287825535:web:98396be2b79a49d0fcc34a",

  measurementId: "G-71MX5H8V0X"

};

// Initialize Firebase

const app = initializeApp(firebaseConfig);
export const storage = getStorage(app);

export default app;
