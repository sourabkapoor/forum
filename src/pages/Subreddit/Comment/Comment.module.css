.comment {
  /* color: white; */
  display: flex;
  column-gap: 8px;
  margin-bottom: 10px;
}

.comment__content {
  display: flex;
  align-items: center;
}

.comment__author {
  margin-bottom: 6px;
  font-size: 12px;
  font-weight: 500;
  line-height: 16px;
}

.comment__date {
  font-size: 9px;
  color: rgb(129, 131, 132);
}

.comment__message {
  font-size: 14px;
  margin-bottom: 6px;
}

.comment__treadline {
  height: 50px;
  border-right: 1px solid #343536;
  display: block;
  padding: 0;
  margin: 0 auto;
  cursor: pointer;
}

.comment__icon {
  width: 36px;
  height: 36px;
  margin: 0 auto;
}
.hide {
  display: none;
}

.comment__profile {
  display: flex;
  align-items: center;
  justify-content: center;
  width: 28px;
  height: 28px;
  border-radius: 50%;
  transform-origin: bottom center;
  /* background-color: #343536; */
  margin-bottom: 12px;
}

.comment__author {
}

.collapse-line {
  border: none;
  background: none;
  padding: 0;
  width: 8px;
  margin-top: 0.5rem;
  position: relative;
  cursor: pointer;
  outline: none;
  transform: translateX(-50%);
}

.collapse-line:hover::before,
.collapse-line:focus-visible::before {
  background-color: hsl(235, 100%, 60%);
}

.collapse-line::before {
  content: "";
  position: absolute;
  top: 0;
  bottom: 0;
  left: 50%;
  width: 1px;
  background-color: hsl(235, 50%, 74%);
  transition: background-color 100ms ease-in-out;
}

.comment__button {
  --hue: 235;
  --color: hsl(var(--hue), 100%, 67%);
  padding: 0.5em 1em;
  margin-top: 1rem;
  background: var(--color);
  border: none;
  color: white;
  border-radius: 0.5em;
  font-size: 0.75em;
  cursor: pointer;
}
