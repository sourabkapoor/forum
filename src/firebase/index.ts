import { initializeApp } from "firebase/app";
import {
  getAuth,
  onAuthStateChanged,
  GoogleAuthProvider,
  signInWithPopup,
  signOut,
  Auth,
} from "firebase/auth";
import {
  getFirestore,
  collection,
  addDoc,
  query,
  orderBy,
  limit,
  onSnapshot,
  setDoc,
  updateDoc,
  doc,
  serverTimestamp,
} from "firebase/firestore";
import {
  getStorage,
  ref,
  uploadBytesResumable,
  getDownloadURL,
} from "firebase/storage";
import { getMessaging, getToken, onMessage } from "firebase/messaging";
import { getPerformance } from "firebase/performance";
import { getAnalytics } from "firebase/analytics";

import app from "./firebase-config";
import { authStateObserver } from "../redux/store";

export const auth = getAuth(app);

export const db = getFirestore(app);

export async function signIn() {
  const provider = new GoogleAuthProvider();
  await signInWithPopup(auth, provider);
}

export async function signOutUser() {
  await signOut(auth);
}

export function initFirebaseAuth() {
  onAuthStateChanged(auth, authStateObserver);
}

export function getUserName() {
  return auth.currentUser?.displayName;
}

export function isUserSignedIn() {
  return !!auth.currentUser;
}

export function getUserId() {
  return auth.currentUser?.uid;
}

export function getUser() {
  return auth.currentUser;
}

initFirebaseAuth();

const analytics = getAnalytics(app);

