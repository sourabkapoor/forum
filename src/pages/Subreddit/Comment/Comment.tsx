import {
  collection,
  doc,
  DocumentData,
  getDocs,
  increment,
  onSnapshot,
  query,
  serverTimestamp,
  setDoc,
  where,
  writeBatch,
} from "firebase/firestore";
import AccountCircleOutlinedIcon from '@mui/icons-material/AccountCircleOutlined';
import moment from "moment";
import { nanoid } from "nanoid";
import React, { useEffect, useState } from "react";
import CSSModules from "react-css-modules";
import {
  selectAuthStatus,
  toggleSignInModal,
} from "../../../features/auth/authSlice";
import { db, getUserId, getUserName } from "../../../firebase";
import { useAppSelector, useAppDispatch } from "../../../hooks/hooks";
import CommentInteractions from "../CommentInteractions/CommentInteractions";
import Comments from "../Comments/Comments";
import styles from "./Comment.module.css";

interface Props {
  comment: DocumentData;
  children: JSX.Element | JSX.Element[];
  postId: string | undefined;
  id: string;
}

const Comment: React.FC<Props> = ({ comment, children, postId, id }) => {
  const [childComments, setChildComments] = useState<
    DocumentData | undefined
  >();
  const [areChildrenHidden, setAreChildrenHidden] = useState(false);

  useEffect(() => {
    async function fetchChildComments() {
      try {
        if (!id) return;
        const childCommentsRef = collection(db, "comments");

        const q = query(childCommentsRef, where("parentId", "==", id));

        onSnapshot(q, (snapshot) => {
          setChildComments(snapshot.docChanges());
        });
      } catch (error) {
        console.log(`Could not fetch child comments: ${error}`);
      }
    }

    fetchChildComments();
  }, [id]);

  return (
    <div styleName="comment">
      <div styleName="comment__user">
        <div styleName="comment__profile">
          <AccountCircleOutlinedIcon />
        </div>
        <button
          styleName="comment__treadline"
          aria-label="Hide Replies"
          onClick={() => setAreChildrenHidden(true)}
        />
      </div>
      <div styleName="comment__content">
        <div styleName="comment__body">
          <div styleName="comment__author">
            <p data-testid="author-description">
              {comment?.userName}{" "}
              <span styleName="comment__date">
                {moment(new Date(comment?.createdAt?.seconds * 1000)).fromNow()}
              </span>
            </p>
          </div>
          <div styleName="comment__message">{comment?.content}</div>
          {children}
          <div styleName={`${areChildrenHidden && "hide"}`}>
            <Comments comments={childComments} commentsPostId={postId} />
          </div>
          <button
            styleName={`comment__button ${!areChildrenHidden ? "hide" : ""}`}
            onClick={() => setAreChildrenHidden(false)}
          >
            Show Replies
          </button>
        </div>
      </div>
    </div>
  );
};

export default CSSModules(Comment, styles, {
  allowMultiple: true,
  handleNotFoundStyleName: "ignore",
});
