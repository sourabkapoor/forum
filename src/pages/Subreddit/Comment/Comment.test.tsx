import React from "react";
import { queryByText, render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import userEvent from "@testing-library/user-event";
import each from "jest-each";
import Comment from "./Comment";
import { getUserName } from "../../../firebase";

describe("Comment", () => {
  test("snapshot", () => {
    const mockProps = {
      comment: {
        content: "this is a mock comment",
      },
      id: "bob",
      postId: "bob jones"
    };
    const { asFragment } = render(
      <Comment {...mockProps}>
        <div>Mock Children</div>
      </Comment>
    );

    expect(asFragment()).toMatchSnapshot();
  });
  test("rendering content correctly", () => {
    const mockProps = {
      comment: {
        content: "this is a mock comment",
        createdAt: {
          seconds: 1662025359,
        },
      },
      id: "bob",
      postId: "bob jones"
    };
    render(
      <Comment {...mockProps}>
        <div>Mock Children</div>
      </Comment>
    );

    const commentText = screen.queryByText("this is a mock comment");

    expect(commentText).toBeInTheDocument();
  });
  test("renders username correctly", () => {
    const mockProps = {
      comment: {
        content: "this is a mock comment",
        createdAt: {
          seconds: 1662025359,
        },
        userName: "bob jones"
      },
      id: "bob",
      postId: "bob jones"
    };
    jest.mock("../../../firebase", () => ({
      ...jest.requireActual("../../../firebase"),
      getUserName: () => "bob jones",
    }));

    render(
      <Comment {...mockProps}>
        <div>Mock Children</div>
      </Comment>
    );

    const userName = screen.queryByTestId("author-description");

    expect(userName?.textContent).toEqual("bob jones")
  });
});
