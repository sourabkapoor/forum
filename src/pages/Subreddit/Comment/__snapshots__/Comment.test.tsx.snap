// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comment snapshot 1`] = `
<DocumentFragment>
  <div
    class="comment"
  >
    <div
      class="comment__user"
    >
      <div
        class="comment__profile"
      >
        <img
          alt="default profile icon"
          class="comment__icon"
          src="https://styles.redditmedia.com/t5_g2km0/styles/profileIcon_snood9e1c00c-b1c6-46a9-a231-a6fcd78cdd16-headshot.png?width=256&height=256&crop=256:256,smart&s=bba24bf813728096bcea480988469df3e9e45c5d"
        />
      </div>
      <div
        class="comment__treadline"
      />
    </div>
    <div
      class="comment__content"
    >
      <div
        class="comment__body"
      >
        <div
          class="comment__author"
        >
          <p
            data-testid="author-description"
          >
            undefined 
            <span
              class="comment__date"
            >
              Invalid date
            </span>
          </p>
        </div>
        <div
          class="comment__message"
        >
          this is a mock comment
        </div>
        <div>
          Mock Children
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
